core/trino-spi/src/main/java/io/trino/spi/connector/MergePage.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi.connector;
+
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+
+import java.util.Optional;
+import java.util.stream.IntStream;
+
+import static io.trino.spi.connector.ConnectorMergeSink.DELETE_OPERATION_NUMBER;
+import static io.trino.spi.connector.ConnectorMergeSink.INSERT_OPERATION_NUMBER;
+import static io.trino.spi.type.TinyintType.TINYINT;
+import static java.lang.Math.toIntExact;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Separate deletions and insertions pages from a merge using
+ * {@link RowChangeParadigm#DELETE_ROW_AND_INSERT_ROW}.
+ */
+public final class MergePage
```

Comment:
This isn't a great name as this is really a merge page for DELETE_ROW_AND_INSERT_ROW style connectors.  I don't think it is worth changing now, but when UPDATE is mapped to merge, mabye we can rename this at the same time

Reply:
Added a TODO to that effect.